src/node_api.cc

```diff
@@ -149,14 +148,19 @@ class ThreadSafeFunction : public node::AsyncResource {
 
   napi_status Push(void* data, napi_threadsafe_function_call_mode mode) {
     node::Mutex::ScopedLock lock(this->mutex);
-    uv_thread_t current_thread = uv_thread_self();
 
     while (queue.size() >= max_queue_size &&
         max_queue_size > 0 &&
         !is_closing) {
       if (mode == napi_tsfn_nonblocking) {
         return napi_queue_full;
-      } else if (uv_thread_equal(&current_thread, &main_thread)) {
+      } else if (v8::Isolate::GetCurrent() != nullptr) {
```

Comment:
There's a reason I didn't suggestion this: it yields false positives with add-ons that create their own isolates on separate threads. (Not a hypothetical, such add-ons exist.)

Incomplete solution: store `env->isolate` in the constructor and check `isolate == v8::Isolate::GetCurrent()`. Incomplete because add-ons can switch isolates on the "home" thread although that _probably_ is a hypothetical.

Reply:
@bnoordhuis @addaleax what about

```C++
v8::Isolate* isolate = v8::Isolate::GetCurrent();
if (isolate != nullptr) {
  node::Environment* node_env = node::Environment::GetCurrent(isolate);
  if (node_env != nullptr && node_env->event_loop() != nullptr) {
    return napi_would_deadlock;
  }
}
```

the reasoning being that we must not `cond->Wait(lock);` on any thread that has one of our event loops? It's really the presence on the thread of a Node.js event loop that precludes blocking here.